sklearn/neighbors/quad_tree.pyx

```diff
@@ -0,0 +1,547 @@
+# cython: boundscheck=False
+# cython: wraparound=False
+# cython: cdivision=True
+# Author: Thomas Moreau <thomas.moreau.2010@gmail.com>
+# Author: Olivier Grisel <olivier.grisel@ensta.fr>
+
+
+from cpython cimport Py_INCREF, PyObject
+
+from libc.stdlib cimport malloc, free
+from libc.string cimport memcpy
+from libc.stdio cimport printf
+
+from sklearn.tree._utils cimport safe_realloc, sizet_ptr_to_ndarray
+from ..utils import check_array
+
+import numpy as np
+cimport numpy as np
+np.import_array()
+
+cdef extern from "math.h":
+    float fabsf(float x) nogil
+
+cdef extern from "numpy/arrayobject.h":
+    object PyArray_NewFromDescr(object subtype, np.dtype descr,
+                                int nd, np.npy_intp* dims,
+                                np.npy_intp* strides,
+                                void* data, int flags, object obj)
+
+
+cdef SIZE_t DEFAULT = <SIZE_t>(-1)
+
+
+# Repeat struct definition for numpy
+CELL_DTYPE = np.dtype({
+    'names': ['parent', 'children', 'cell_id', 'point_index', 'is_leaf',
+              'max_width', 'depth', 'cumulative_size', 'center', 'barycenter',
+              'min_bounds', 'max_bounds'],
+    'formats': [np.intp, (np.intp, 8), np.intp, np.intp, np.int32, np.float32, 
+                np.intp, np.intp, (np.float32, 3), (np.float32, 3),
+                (np.float32, 3), (np.float32, 3)],
+    'offsets': [
+        <Py_ssize_t> &(<Cell*> NULL).parent,
+        <Py_ssize_t> &(<Cell*> NULL).children,
+        <Py_ssize_t> &(<Cell*> NULL).cell_id,
+        <Py_ssize_t> &(<Cell*> NULL).point_index,
+        <Py_ssize_t> &(<Cell*> NULL).is_leaf,
+        <Py_ssize_t> &(<Cell*> NULL).max_width,
+        <Py_ssize_t> &(<Cell*> NULL).depth,
+        <Py_ssize_t> &(<Cell*> NULL).cumulative_size,
+        <Py_ssize_t> &(<Cell*> NULL).center,
+        <Py_ssize_t> &(<Cell*> NULL).barycenter,
+        <Py_ssize_t> &(<Cell*> NULL).min_bounds,
+        <Py_ssize_t> &(<Cell*> NULL).max_bounds,
+    ]
+})
+
+assert CELL_DTYPE.itemsize == sizeof(Cell)
+
+
+cdef class QuadTree:
+    """Array-based representation of a QuadTree.
+    """
+    def __cinit__(self, int n_dimensions, int verbose):
+        """Constructor."""
+        # Parameters of the tree
+        self.n_dimensions = n_dimensions
+        self.verbose = verbose
+        self.n_cells_per_cell = 2 ** self.n_dimensions
+
+        # Inner structures
+        self.max_depth = 0
+        self.cell_count = 0
+        self.capacity = 0
+        self.n_points = 0
+        self.cells = NULL
+
+    def __dealloc__(self):
+        """Destructor."""
+        # Free all inner structures
+        free(self.cells)
+
+    property cumulative_size:
+        def __get__(self):
+            return self._get_cell_ndarray()['cumulative_size'][:self.cell_count]
+
+    property leafs:
+        def __get__(self):
+            return self._get_cell_ndarray()['is_leaf'][:self.cell_count]
+
```

Comment:
The properties give easy access to properties of all the cells in python by giving back the property in a single numpy array.  
These 2 are used in the tests but I can add other cell properties if needed.